src/platform/qt/library/LibraryModel.cpp

```diff
@@ -0,0 +1,417 @@
+		spanSet.sort(true);
+		for (const SpanSet::Span& span : spanSet.spans) {
+			if (m_treeMode) {
+				beginRemoveRows(parent, span.left, span.right);
+			}
+			pathIndex.erase(pathIndex.begin() + span.left, pathIndex.begin() + span.right + 1);
+			if (m_treeMode) {
+				endRemoveRows();
+			}
+		}
+	}
+	removedRootSpans.merge();
+	removedRootSpans.sort(true);
+	for (const SpanSet::Span& span : removedRootSpans.spans) {
+		beginRemoveRows(QModelIndex(), span.left, span.right);
+		if (m_treeMode) {
+			for (int i = span.right; i >= span.left; i--) {
+				QString base = m_pathOrder.takeAt(i);
+				m_pathIndex.remove(base);
+			}
+		} else {
+			m_games.erase(m_games.begin() + span.left, m_games.begin() + span.right + 1);
+		}
+		endRemoveRows();
+	}
+	for (int i = m_games.size() - 1; i >= firstModifiedIndex; i--) {
+		m_gameIndex[m_games[i]->fullpath] = i;
+	}
+}
+
+QModelIndex LibraryModel::index(const QString& game) const {
+	int pos = m_gameIndex.value(game, -1);
+	if (pos < 0) {
+		return QModelIndex();
+	}
+	if (m_treeMode) {
+		const LibraryEntry& entry = *m_games[pos];
+		return createIndex(m_pathIndex[entry.base].indexOf(&entry), 0, m_pathOrder.indexOf(entry.base));
+	}
+	return createIndex(pos, 0);
+}
+
+QModelIndex LibraryModel::index(int row, int column, const QModelIndex& parent) const {
+	if (!parent.isValid()) {
+		return createIndex(row, column, quintptr(0));
+	}
+	if (!m_treeMode || parent.internalId() || parent.column() != 0) {
+		return QModelIndex();
+	}
+	return createIndex(row, column, parent.row() + 1);
+}
+
+QModelIndex LibraryModel::parent(const QModelIndex& child) const {
+	if (!child.isValid() || child.internalId() == 0) {
+		return QModelIndex();
+	}
+	return createIndex(child.internalId() - 1, 0, quintptr(0));
+}
+
+int LibraryModel::columnCount(const QModelIndex& parent) const {
+	if (!parent.isValid() || (parent.column() == 0 && !parent.parent().isValid())) {
+		return MAX_COLUMN + 1;
+	}
+	return 0;
+}
+
+int LibraryModel::rowCount(const QModelIndex& parent) const {
+	if (parent.isValid()) {
+		if (m_treeMode) {
+			if (parent.row() < 0 || parent.row() >= m_pathOrder.size() || parent.column() != 0) {
+				return 0;
+			}
+			return m_pathIndex[m_pathOrder[parent.row()]].size();
+		}
+		return 0;
+	}
+	if (m_treeMode) {
+		return m_pathOrder.size();
+	}
+	return m_games.size();
+}
+
+QVariant LibraryModel::folderData(const QModelIndex& index, int role) const
+{
+	// Precondition: index and role must have already been validated
+	if (role == Qt::DecorationRole) {
+		return qApp->style()->standardIcon(QStyle::SP_DirOpenIcon);
+	}
+	if (role == FullPathRole || (index.column() == COL_LOCATION && role != Qt::DisplayRole)) {
+		return m_pathOrder[index.row()];
+	}
+	if (index.column() == COL_NAME) {
+		QString path = m_pathOrder[index.row()];
+		return path.section('/', -1);
+	}
+	return QVariant();
+}
+
+QVariant LibraryModel::data(const QModelIndex& index, int role) const {
+	if (role != Qt::DisplayRole &&
```

Comment:
I think it might be a bit less repetitive to turn this into a switch/case with default being the early return and the valid fall-through cases just being a break statement.